add device side `ForEachParticle`

psychocoderHPC opened the pull request:
Add device-side `ForEachParticle` algorithm.

- add a policy to describe if particle frames should be forward or backward processed
- add factory functions to create the particle forEach executor.
- The particle for each is providing two modes
  - iterate over all particles in a supercell
  - iterate over all frames -> required mostly for pmacc internal algorithms to compress the linked frame data structures.

Depending on the number of worker and the frame size the forach algorithm will work with skip lists to utilize all threads.

psychocoderHPC committed:
add device side ForEachParticles algorithm

- add policy to describe if particle frames should forward or backward processed
- add factory functions to create the particle forEach executor.
- The particle for each is providing two modes
  - iterate over all particles in a supercell
  - iterate over all frames -> required mostly for pmacc internal algorithms to compress the linked frame data structures.

psychocoderHPC committed:
use device side FrameForEach

psychocoderHPC committed:
use device side ForEachParticle

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+            template<typename T_ForEach>
+            static DINLINE auto createCtx(T_ForEach& forEachParticleInFrame)
+            {
+                return lockstep::makeVar<lcellId_t>(forEachParticleInFrame, lcellId_t(0u));
```
It is different from `Forward<>::createCtx()` on purpose? Here you use `lcellId_t`, and there `uint32_t`.

psychocoderHPC replied:
Yes depending on the iteration order the variable is used differently. `lcellId_t` is enough to store the number of particles in a frame.
For the other directing the variable is used to store number of particles in a supercell, there it mus the 32Bit.

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
-                            // offset of the superCell (in cells, without any guards) to the origin of the local domain
-                            DataSpace<dim> const localSuperCellOffset = superCellIdx - mapper.getGuardingSuperCells();
+            using SuperCellSize = typename T_ParBox::FrameType::SuperCellSize;
+            static constexpr uint32_t frameSize = pmacc::math::CT::volume<SuperCellSize>::type::value;
```
I think this may be a good moment to separate frame size from supercell size, e.g. in a follow-up PR. Could start as simple as a function that always returns `pmacc::math::CT::volume<SuperCellSize>::type::value` that we use now (it's not just here, but basically everywhere in the code).

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
-                                = functor(acc, localSuperCellOffset, lockstep::Worker<T_numWorkers>{workerIdx});
+            /** Number of frames to skip.
+             *
+             * If we have more worker than particles in a frame we skip frames depending on the worker id.
```
Typo: should be `workers`

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+             *
+             * If we have more worker than particles in a frame we skip frames depending on the worker id.
+             */
+            static constexpr uint32_t numFramesToSkip = (numWorkers + frameSize - 1u) / frameSize;
```
I think this is confusing - does this variable actually mean the number of frames to skip + 1?  E.g. for `numWorkers == frameSize` its value will be 1, while we should skip 0 frames in this case.

(Same for the twin of this variable inside the lambda)

psychocoderHPC replied:
You are right the distance to the next frame. Let's check this tomorrow offline and find a better naming.

psychocoderHPC replied:
I will change it to `frameDistance`

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+         * @tparam T_numWorkers Number of lockstep workers.
+         */
+        template<typename T_AccessTag, typename T_Order, typename T_ParBox, uint32_t T_numWorkers>
+        struct ForEachParticle
```
Matter of preference, but I would make this a class given it has a private state

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
              *
-             * Has a version for a fixed area, and for a user-provided mapper factory.
-             * They differ only in how the area is defined.
+             * Depending of the number of worker used it could be that more than one frame executed by all workers in
```
Typo: should be `workers`

sbastrakov replied:
Another typo: instead of `frame executed` should be `frame is executed`

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+             * @param particlesBox particles memory
+             *                     It is not allowed concurrently to add or remove particles during the
+             *                     execution of this algorithm.
+             * @param superCellIdx index of the superCell where particles should be processed
```
Is it really a supercell index, or more generally an index inside the given particle box?

If the latter, I think it is better to rename in order to separate the basic for each logic from the normal PIConGPU way of organizing parallel particle processing

sbastrakov replied:
Perhaps a related thought - do we need to operate as the whole box and supercell index here? It would appear more logical to have a single supercell already as our collection of particles to operate on. As then we would have a more crisp and standard parallel for-each logic here.

A user interface could still be a box + index if that is more convenient, a transition could also be made inside

psychocoderHPC replied:
It is supercell index we need to operate on.

>  do we need to operate as the whole box and supercell index here? 

I also thought first of it but because of how pmacc is structured internally working with a supercell will not be easily possible without other refactorings in pmacc. There is no link from a supercell to the particle box you need to iterate over frames, therefore I decided to work with the supercell index.

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
-             *                             adheres to the AreaMapperFactory concept
+             * @attention There is no guarantee in which order particles will be executed.
+             *            If the particle functor is used the algorithm assumes that the frame structure is following
+             *            the rule tat all except the last frame are fully filled and in the last frame the low indices
```
Typo: should be `that`

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
-             *                             adheres to the AreaMapperFactory concept
+             * @attention There is no guarantee in which order particles will be executed.
+             *            If the particle functor is used the algorithm assumes that the frame structure is following
+             *            the rule tat all except the last frame are fully filled and in the last frame the low indices
+             *            within the frames will be contigious filled too.
```
Would better be pharsed as
`the rule that all frames but the last are fully filled and in the last frame the low indices are contigiously filled.`

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+             *                     different workers is data race free.
+             *                     particle functor: 'void operator()(T_Acc const &, ParticleType)'.
+             *                       It is not allowed to call a synchronization function within the functor.
+             *                     frame functor: 'void operator()(T_Acc const &, FrameCtx)'.
```
Did you mean `For frame ...` ?

psychocoderHPC replied:
I will put a dash in front to separate that we handle once a particle functor and once the frame functor.

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+            template<typename T_Acc, typename T_Functor>
+            DINLINE void operator()(T_Acc const& acc, T_Functor&& unaryFunctor) const
+            {
+                if constexpr(numWorkers > frameSize)
+                {
+                    PMACC_CASSERT_MSG(
+                        __number_of_workers_must_be_a_multiple_of_framesize,
+                        (numWorkers % frameSize) == 0);
+                }
```
The error message does not match the condition: from the condition `numWorkers` can be either `<= frameSize` or a multiple of it

psychocoderHPC replied:
All looks fine, in cases where we have more workers than particles in a frame the number of workers must be a multiple of the frame size.

sbastrakov replied:
So the condition is fine. But then the name does not exactly express the condition. We can discuss tomorrow

psychocoderHPC replied:
should be `__number_of_workers_must_be_either_less_or_a_multiple_of_framesize`

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+                // Information used to know if a particle within a frame is active without checking the multiMask.
+                auto frameConditionData = T_Order::createCtx(forEachParticleInFrame);
+
+                // end kernel if we have no particles
+                if(numPartcilesInSupercell == 0)
+                    return;
```
I think this check is better put right after we got `numPartcilesInSupercell`. Or should `frameConditionData` be created always - if so, please add a comment about it?

psychocoderHPC replied:
I can move `frameConditionData` down.

sbastrakov replied:
So `createCtx` is not a collective operation?

psychocoderHPC replied:
It is a collective operation but is not requiring any synchronization.

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+                }
+
+                auto const& superCell = m_particlesBox.getSuperCell(m_superCellIdx);
+                uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
```
Typo: should be `numParticlesInSupercell`

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+                        {
+                            if(frame.isValid())
+                                frameConditionData[idx]
+                                    = m_particlesBox.getSuperCell(m_superCellIdx).getSizeLastFrame();
```
Can variable `superCell` be used instead?

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+                        return frame;
+                    });
+
+                bool isOneFrameValid = true;
```
Why `one`? To me it seems more like a `isCurrentFrameValid`

psychocoderHPC replied:
A worker can have multiple frames depending on the number of virtual threads the worker is representing.
We need to know if one of these frames a worker is holding is valid.

psychocoderHPC replied:
offline discussed @sbastrakov suggested: `hasUnprocessedFrames` or `hasMoreWork`

sbastrakov commented on include/picongpu/particles/debyeLength/Estimate.kernel:
```diff
@@ -118,28 +118,24 @@ namespace picongpu
-                        frame = electronBox.getPreviousFrame(frame);
-                        particlesInFrame = frameSize;
-                    }
+                    auto forEachParticle = pmacc::particles::algorithm::acc::makeForEach<numWorkers>(
+                        workerIdx,
+                        electronBox,
+                        supercellIdx);
```
As a idea for a follow-up PR (this one is already very large).

With these changes, I think we can finally go away from operating directly with `numWorkers` and `workerIdx` in "simple" kernels that just need to go over the particles. With shared memory like for current deposition I understand those are still needed as this point doesn't apply.

So perhaps we can make a simplified way to create an in-kernel for each particle that uses default `numWorkers`. This way those kernels would not need this (potentially confusing) template parameter or directly use thread indices. Ultimately, then such kernels would become a more direct for-each invocation as they imo should.

psychocoderHPC replied:
This is only possible if forEachParticle is rewritten to work with runtime sizes only. The number of workers depends on the alpaka accelerator we are using. I would avoid using any `default numWorkers` because the default will never be good. 
A forEach which is fully runtime would IMO make more sense.

sbastrakov replied:
What i meant is, most (not sure, maybe some, not most) PIConGPU particle-based kernels (current deposition is exception) use template parameter `uint32_t T_numWorkers` and on the host side we use `pmacc::traits::GetNumWorkers<...>` to provide a value for it. I was thinking to just simplify this common case by using this value by default.

sbastrakov replied:
Discussed offline, as a follow-up to this work we can make some helpers to simplify writing and launching such kernels. That may also include ensuring that the `T_numWorkers` (or a new way to express it) matches the kernel block size - which is a requirement even now, but we do not check for it.

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+
+                        if constexpr(std::is_same_v<Reverse, T_Order>)
+                        {
+                            if(frame.isValid())
```
Why is this condition needed? If we had no particles we would have returned already

sbastrakov replied:
@psychocoderHPC what is your opinion on this?

psychocoderHPC replied:
Yes I forgot about it. It is not necessarily required.
If there are no must-have changes I will fix it in a separate PR.

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+                uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+
+                // Information used to know if a particle within a frame is active without checking the multiMask.
+                auto frameConditionData = T_Order::createCtx(forEachParticleInFrame);
```
I think this variable is very confusingly named. Need to discuss on what does it express

sbastrakov replied:
I still strongly dislike the name, but can live with it. After offline discussion it seems the concept represented by this variable is also quite confusing but has to be so due to performance and register reasons (and this is definitely a performance-critical bit).

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+                                    if(!isParticle)
+                                        particle.setHandleInvalid();
+                                }
+                                else
+                                    static_assert(!sizeof(T_Order), "Unsupported order policy");
```
Discussed offline: could be kept as-is, but may be moved somewhere (e.g. checked even outside this function), as it is a precondition on the class template argument.

sbastrakov commented on include/pmacc/particles/algorithm/ForEach.hpp:
```diff
@@ -22,157 +22,447 @@
+                        constexpr uint32_t skipNFrames = (numWorkers + frameSize - 1u) / frameSize;
+                        if constexpr(skipNFrames > 1)
+                        {
+                            uint32_t const frameIdx = idx / frameSize;
```
This  variable looks more like the number of skipped frames

psychocoderHPC committed:
workaround g++7 bug

g++7 with alpaka OpenMP2 accelerator is showing the following error:

```
pmacc/particles/ParticlesBase.kernel:515:55: internal compiler error: in finish_member_declaration, at cp/semantics.c:3029
                             return isValidSupercellIdx(neighborSuperCellIdx, numSupercellsWithGuard);
                                    ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
```

psychocoderHPC commented:
The CI should now hopefully pass all tests.
